src/Base/PyWrapParseTupleAndKeywords.h

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+/****************************************************************************
+ *                                                                          *
+ *   Copyright (c) 2023 The FreeCAD Project Association                     *
+ *                                                                          *
+ *   This file is part of FreeCAD.                                          *
+ *                                                                          *
+ *   FreeCAD is free software: you can redistribute it and/or modify it     *
+ *   under the terms of the GNU Lesser General Public License as            *
+ *   published by the Free Software Foundation, either version 2.1 of the   *
+ *   License, or (at your option) any later version.                        *
+ *                                                                          *
+ *   FreeCAD is distributed in the hope that it will be useful, but         *
+ *   WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       *
+ *   Lesser General Public License for more details.                        *
+ *                                                                          *
+ *   You should have received a copy of the GNU Lesser General Public       *
+ *   License along with FreeCAD. If not, see                                *
+ *   <https://www.gnu.org/licenses/>.                                       *
+ *                                                                          *
+ ***************************************************************************/
+
+#ifndef FREECAD_PYWRAPPARSETUPLEANDKEYWORDS_H
+#define FREECAD_PYWRAPPARSETUPLEANDKEYWORDS_H
+
+#include <Python.h>
+#include <array>
+#include <type_traits>
+
+namespace Base {
+
+    /// A thin C++ wrapper around PyArg_ParseTupleAndKeywords providing const-correctness for the keywords argument
+    /// \arg args The Python args object
+    /// \arg kw The Python kw object
+    /// \arg format The format string describing the expected arguments
+    /// \arg keywords A std::array of keywords, terminated by a nullptr (required by CPython)
+    /// \arg (variadic) Pointers to the storage locations for the parameters
+    /// \returns boolean true on success, or false on failure
+    template<size_t arraySize>
+    bool Wrapped_ParseTupleAndKeywords(PyObject *args,
+                                       PyObject *kw,
+                                       const char *format,
+                                       const std::array<const char *, arraySize> keywords,
+                                       ...)
+    {
+        // NOTE: This code is from getargs.c in the Python source code (modified to use the public interface at
```

Comment:
Maybe use a `static_assert(arraySize > 0, "Array must at least have a single element");` to guarantee that no empty array is passed.

Reply:
Good idea, I've done so now. I'd love to also do a compile-time check on the last argument being a `nullptr`, but my metaprogramming skills are too weak.

Reply:
> I'd love to also do a compile-time check on the last array entry being a nullptr, but my metaprogramming skills are too weak.

I don't think this is possible because how could the function make guarantees at compile time about the std::array it gets from elsewhere? All what you can do is to make this guarantee in the calling instance:

```
constexpr std::array<const char *, 3> keywords = {"1", "2", nullptr};
static_assert(keywords.back() == nullptr, "Last element must be a nullptr");

Wrapped_ParseTupleAndKeywords(args, kwds, "|i", keywords, &compression)
...
```
But having the static_assert at the some scope where the std::array is created is pointless, IMO.

Actually, it is sufficient that this check is done at runtime and in case it fails you should set a Python exception and return false. This way the exception handling of the calling instances doesn't need to be changed.
```
bool Wrapped_ParseTupleAndKeywords(PyObject *args,
                                       PyObject *kw,
                                       const char *format,
                                       const std::array<const char *, arraySize> keywords,
                                       ...)
{
...
    if (keywords.back()) {
        PyErr_SetString(PyExc_ValueError, "Last element of keywords array is not null");
        return false;
    }
...
}
```

Reply:
> I don't think this is possible because how could the function make guarantees at compile time about the std::array it gets from elsewhere?

You're right -- metaprogramming is definitely not my strength!

> But having the static_assert at the some scope where the std::array is created is pointless, IMO.

Agree completely.

I've added a commit that implements your suggestion for making the error message better at runtime.